src/meepgeom.cpp

```diff
@@ -1145,6 +1188,62 @@ void geom_epsilon::eff_chi1inv_matrix(meep::component c, symmetric_matrix *chi1i
 static int eps_ever_negative = 0;
 static meep::field_type func_ft = meep::E_stuff;
 
+struct matgrid_volavg {
+  meep::ndim dim;        // dimensionality of voxel
+  double rad;            // (spherical) voxel radius
+  double uval;           // bilinearly-interpolated weight at voxel center
+  double ugrad_abs;      // magnitude of gradient of uval
+  double beta;           // thresholding bias
+  double eta;            // thresholding erosion/dilation
+  double eps1;           // trace of epsilon tensor from medium 1
+  double eps2;           // trace of epsilon tensor from medium 2
+};
+
+static void get_uproj_w(const void *mgva_, double x0, double &uproj_, double &w_) {
+  matgrid_volavg *mgva = (matgrid_volavg *)mgva_;
+  // use a linear approximation for the material grid weights around the Yee grid point
+  uproj_ = tanh_projection(mgva->uval + mgva->ugrad_abs*x0, mgva->beta, mgva->eta);
+  if (mgva->dim == meep::D1)
+    w_ = 1/(2*mgva->rad);
+  else if (mgva->dim == meep::D2)
+    w_ = 2*sqrt(mgva->rad*mgva->rad - x0*x0)/(meep::pi * mgva->rad*mgva->rad);
+  else if (mgva->dim == meep::D3)
+    w_ = meep::pi*(mgva->rad*mgva->rad - x0*x0)/(4/3 * meep::pi * mgva->rad*mgva->rad*mgva->rad);
+}
+
+#ifdef CTL_HAS_COMPLEX_INTEGRATION
+static cnumber matgrid_ceps_func(int n, number *x, void *mgva_) {
+  double u_proj = 0, w = 0;
+  get_uproj_w(mgva_, x[0], u_proj, w);
+  matgrid_volavg *mgva = (matgrid_volavg *)mgva_;
+  double eps1 = mgva->eps1;
+  double eps2 = mgva->eps2;
+  cnumber ret;
+  ret.re = (1-u_proj)*eps1 + u_proj*eps2;
+  ret.im = (1-u_proj)/eps1 + u_proj/eps2;
+  return ret * w;
+}
+#else
+static number matgrid_eps_func(int n, number *x, void *mgva_) {
+  double u_proj = 0, w = 0;
+  get_uproj_w(mgva_, x[0], u_proj, w);
+  matgrid_volavg *mgva = (matgrid_volavg *)mgva_;
+  double eps1 = mgva->eps1;
+  double eps2 = mgva->eps2;
+  double eps_interp = (1-u_proj)*eps1 + u_proj*eps2;
+  return eps_interp * w;
```

Comment:
You might as well combine these 4 lines:
```cxx
return w * ((1-u_proj)*mgva->eps1 + u_proj*mgva->eps2);
```

Reply:
Fixed.